macOS build is broken

The bundling fails with:

```
+ cpack
CPack: Create package using DragNDrop
CPack: Install projects
CPack: - Run preinstall target for: copyq
CPack: - Install project: copyq []
CMake Error at /usr/local/Cellar/cmake/3.27.4/share/cmake/Modules/BundleUtilities.cmake:458 (message):
  otool -l failed: 1

  error:
  /Applications/Xcode_13.2.1.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/otool-classic:
  can't open file:
  @loader_path/../../../../../opt/freetype/lib/libfreetype.6.dylib (No such
  file or directory)

Call Stack (most recent call first):
  /usr/local/Cellar/cmake/3.27.4/share/cmake/Modules/BundleUtilities.cmake:527 (get_item_rpaths)
  /usr/local/Cellar/cmake/3.27.4/share/cmake/Modules/BundleUtilities.cmake:614 (set_bundle_key_values)
  /usr/local/Cellar/cmake/3.27.4/share/cmake/Modules/BundleUtilities.cmake:933 (get_bundle_keys)
  /Users/runner/work/CopyQ/build/src/cmake_install.cmake:190 (fixup_bundle)
  /Users/runner/work/CopyQ/build/cmake_install.cmake:62 (include)
```

I tried few workarounds but none worked:
- `brew unlink freetype && brew link freetype`
- Use older version of qt@5 - this would need much more work then just using the older `qt@5.rb`
- Move to Qt 6 - KDE Framework still does not support it

--- a/.github/workflows/build-linux.yml
+++ b/.github/workflows/build-linux.yml
@@ -14,7 +14,6 @@ jobs:
         include:
           - os: ubuntu-20.04
             buildname: Linux-GCC
-            build_type: Debug
             compiler: g++
             coverage: true
             compiler_flags: >-
@@ -27,14 +26,12 @@ jobs:
 
           - os: ubuntu-20.04
             buildname: Linux-Clang
-            build_type: Debug
             compiler: clang++
             compiler_package: clang
             with_qt6: false
 
           - os: ubuntu-latest
             buildname: Qt 6
-            build_type: Debug
             compiler: g++
             compiler_package: g++
             with_qt6: true
@@ -45,6 +42,15 @@ jobs:
         with:
           submodules: false
           fetch-depth: 1
+
+      - name: Enable ccache
+        uses: hendrikmuhs/ccache-action@v1.2
+        with:
+          key: ${{ github.job }}-${{ matrix.os }}
+
+      - name: Set up ccache
+        run: |
+          export PATH="/usr/lib/ccache:/usr/local/opt/ccache/libexec:$PATH"
 
       - name: Install dependencies
         uses: nick-invision/retry@v2
@@ -59,25 +65,17 @@ jobs:
             '${{matrix.compiler_package}}'
 
       - name: Build with CMake
-        uses: lukka/run-cmake@v3
+        uses: lukka/run-cmake@v10
         with:
-          cmakeListsOrSettingsJson: CMakeListsTxtAdvanced
-          cmakeListsTxtPath: '${{github.workspace}}/CMakeLists.txt'
-          cmakeBuildType: Release
-          cmakeGenerator: Ninja
-          buildDirectory: '${{runner.workspace}}/build'
-          cmakeAppendedArgs: >-
-            -DWITH_TESTS=TRUE
-            -DPEDANTIC=TRUE
-            -DCMAKE_BUILD_TYPE='${{matrix.build_type}}'
-            -DCMAKE_INSTALL_PREFIX='${{runner.workspace}}/install'
-            -DCMAKE_CXX_FLAGS='${{matrix.compiler_flags}}'
-            -DCMAKE_C_FLAGS='${{matrix.compiler_flags}}'
-            -DCMAKE_CXX_COMPILER='${{matrix.compiler}}'
-            -DWITH_QT6='${{matrix.with_qt6}}'
-
-      - name: Install with CMake
-        run: cmake --install '${{runner.workspace}}/build'
+          configurePreset: Debug
+          buildPreset: Debug
+          configurePresetAdditionalArgs: >-
+            [
+            '-DCMAKE_CXX_COMPILER=${{matrix.compiler}}',
+            '-DCMAKE_CXX_FLAGS=${{matrix.compiler_flags}}',
+            '-DCMAKE_C_FLAGS=${{matrix.compiler_flags}}',
+            '-DWITH_QT6=${{matrix.with_qt6}}'
+            ]
 
       - name: Create gnupg directory for tests
         run: mkdir -p ~/.gnupg

--- a/.github/workflows/build-macos.yml
+++ b/.github/workflows/build-macos.yml
@@ -14,8 +14,6 @@ jobs:
         include:
           - os: macos-11
             buildname: macOS 10.15
-            build_type: Release
-            compiler: clang_64
             create_bundle: true
 
     steps:
@@ -25,23 +23,27 @@ jobs:
           submodules: false
           fetch-depth: 1
 
+      - name: Enable ccache
+        uses: hendrikmuhs/ccache-action@v1.2
+        with:
+          key: ${{ github.job }}-${{ matrix.os }}
+
+      - name: Set up ccache
+        run: |
+          export PATH="/usr/lib/ccache:/usr/local/opt/ccache/libexec:$PATH"
+
       - name: Install dependencies
+        env:
+          HOMEBREW_NO_INSTALLED_DEPENDENTS_CHECK: 1
+          HOMEBREW_NO_AUTO_UPDATE: 1
+          HOMEBREW_NO_INSTALL_CLEANUP: 1
         run: '${{github.workspace}}/utils/github/install-macos.sh'
 
       - name: Build with CMake
-        uses: lukka/run-cmake@v3
+        uses: lukka/run-cmake@v10
         with:
-          cmakeListsOrSettingsJson: CMakeListsTxtAdvanced
-          cmakeListsTxtPath: '${{github.workspace}}/CMakeLists.txt'
-          cmakeBuildType: Release
-          cmakeGenerator: Ninja
-          buildDirectory: '${{runner.workspace}}/build'
-          cmakeAppendedArgs: >-
-            -DWITH_TESTS=TRUE
-            -DCMAKE_BUILD_TYPE='${{matrix.build_type}}'
-            -DCMAKE_PREFIX_PATH="/usr/local/opt/qt@5/lib/cmake"
-            -DCMAKE_OSX_DEPLOYMENT_TARGET='10.15'
-            -DCMAKE_CXX_FLAGS="-Wno-deprecated-declarations"
+          configurePreset: macOS
+          buildPreset: macOS
 
       - name: Create gnupg directory for tests
         run: mkdir -p ~/.gnupg

--- a/CMakePresets.json
+++ b/CMakePresets.json
@@ -1,0 +1,45 @@
+{
+    "version": 3,
+    "configurePresets": [
+        {
+            "name": "Debug",
+            "generator": "Ninja",
+            "binaryDir": "${sourceParentDir}/build",
+            "installDir": "${sourceParentDir}/install",
+            "cacheVariables": {
+                "WITH_TESTS": "TRUE",
+                "PEDANTIC": "TRUE",
+                "CMAKE_C_COMPILER_LAUNCHER": "ccache",
+                "CMAKE_CXX_COMPILER_LAUNCHER": "ccache"
+            }
+        },
+        {
+            "name": "macOS",
+            "generator": "Ninja",
+            "binaryDir": "${sourceParentDir}/build",
+            "installDir": "${sourceParentDir}/build",
+            "cacheVariables": {
+                "WITH_TESTS": "TRUE",
+                "CMAKE_PREFIX_PATH": "/usr/local/opt/qt@5/lib/cmake",
+                "CMAKE_OSX_DEPLOYMENT_TARGET": "10.15",
+                "CMAKE_C_COMPILER_LAUNCHER": "ccache",
+                "CMAKE_CXX_COMPILER_LAUNCHER": "ccache",
+                "CMAKE_CXX_FLAGS": "-Wno-deprecated-declarations"
+            }
+        }
+    ],
+    "buildPresets": [
+        {
+            "name": "Debug",
+            "configurePreset": "Debug",
+            "configuration": "Debug",
+            "targets": "install"
+        },
+        {
+            "name": "macOS",
+            "configurePreset": "macOS",
+            "configuration": "Release",
+            "targets": "install"
+        }
+    ]
+}

--- a/docs/build-source-code.rst
+++ b/docs/build-source-code.rst
@@ -144,7 +144,15 @@ Build with the following commands:
 
     cmake -DCMAKE_PREFIX_PATH="$(brew --prefix qt5)" .
     cmake --build .
-    cpack
+    cmake --install .
+    macdeployqt CopyQ.app -dmg -verbose=2 -always-overwrite \
+        -executable=CopyQ.app/Contents/PlugIns/copyq/libitemfakevim.so \
+        -executable=CopyQ.app/Contents/PlugIns/copyq/libitemimage.so \
+        -executable=CopyQ.app/Contents/PlugIns/copyq/libitemnotes.so \
+        -executable=CopyQ.app/Contents/PlugIns/copyq/libitempinned.so \
+        -executable=CopyQ.app/Contents/PlugIns/copyq/libitemsync.so \
+        -executable=CopyQ.app/Contents/PlugIns/copyq/libitemtags.so \
+        -executable=CopyQ.app/Contents/PlugIns/copyq/libitemtext.so
 
 This will produce a self-contained application bundle ``CopyQ.app``
 which can then be copied or moved into ``/Applications``.

--- a/plugins/itemimage/itemimage.cpp
+++ b/plugins/itemimage/itemimage.cpp
@@ -8,6 +8,10 @@
 #include "item/itemeditor.h"
 #include "gui/pixelratio.h"
 
+#ifdef HAS_TESTS
+#   include "tests/itemimagetests.h"
+#endif
+
 #include <QBuffer>
 #include <QHBoxLayout>
 #include <QModelIndex>
@@ -217,6 +221,17 @@ QStringList ItemImageLoader::formatsToSave() const
         QLatin1String("image/png"),
         QLatin1String("image/gif")
     };
+}
+
+QObject *ItemImageLoader::tests(const TestInterfacePtr &test) const
+{
+#ifdef HAS_TESTS
+    QObject *tests = new ItemImageTests(test);
+    return tests;
+#else
+    Q_UNUSED(test)
+    return nullptr;
+#endif
 }
 
 void ItemImageLoader::applySettings(QSettings &settings)

--- a/plugins/itemimage/itemimage.h
+++ b/plugins/itemimage/itemimage.h
@@ -68,6 +68,8 @@ public:
 
     QStringList formatsToSave() const override;
 
+    QObject *tests(const TestInterfacePtr &test) const override;
+
     void applySettings(QSettings &settings) override;
 
     void loadSettings(const QSettings &settings) override;

--- a/plugins/itemimage/tests/itemimagetests.cpp
+++ b/plugins/itemimage/tests/itemimagetests.cpp
@@ -1,0 +1,112 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+#include "itemimagetests.h"
+
+#include "tests/test_utils.h"
+
+#include <QBuffer>
+#include <QImageReader>
+#include <QImageWriter>
+#include <QImage>
+
+ItemImageTests::ItemImageTests(const TestInterfacePtr &test, QObject *parent)
+    : QObject(parent)
+    , m_test(test)
+{
+}
+
+void ItemImageTests::initTestCase()
+{
+    TEST(m_test->initTestCase());
+}
+
+void ItemImageTests::cleanupTestCase()
+{
+    TEST(m_test->cleanupTestCase());
+}
+
+void ItemImageTests::init()
+{
+    TEST(m_test->init());
+}
+
+void ItemImageTests::cleanup()
+{
+    TEST( m_test->cleanup() );
+}
+
+void ItemImageTests::supportedFormats()
+{
+    const auto supportedRead = QImageReader::supportedImageFormats();
+    QVERIFY(supportedRead.contains("png"));
+    QVERIFY(supportedRead.contains("bmp"));
+
+#if defined(Q_OS_WIN) || defined(Q_OS_MAC)
+    QVERIFY(supportedRead.contains("gif"));
+    QVERIFY(supportedRead.contains("webp"));
+#endif
+
+    const auto supportedWrite = QImageWriter::supportedImageFormats();
+    QVERIFY(supportedWrite.contains("png"));
+    QVERIFY(supportedWrite.contains("bmp"));
+}
+
+void ItemImageTests::savePng()
+{
+    QImage image(":images/icon");
+    QVERIFY(!image.isNull());
+    image = image.scaled(QSize(16, 16));
+
+    QByteArray data;
+    QBuffer buffer(&data);
+    buffer.open(QBuffer::ReadWrite);
+
+    QVERIFY(image.save(&buffer, "PNG"));
+    TEST( m_test->setClipboard("TEST") );
+    WAIT_ON_OUTPUT("read(0)", "TEST");
+    TEST( m_test->setClipboard(data, "image/png") );
+    WAIT_ON_OUTPUT("read('image/png', 0).length > 0", "true\n");
+}
+
+void ItemImageTests::saveBmp()
+{
+    QImage image(":images/icon");
+    QVERIFY(!image.isNull());
+    image = image.scaled(QSize(16, 16));
+
+    QByteArray data;
+    QBuffer buffer(&data);
+    buffer.open(QBuffer::ReadWrite);
+
+    QVERIFY(image.save(&buffer, "BMP"));
+    TEST( m_test->setClipboard("TEST") );
+    WAIT_ON_OUTPUT("read(0)", "TEST");
+    TEST( m_test->setClipboard(data, "image/bmp") );
+    WAIT_ON_OUTPUT("read('image/png', 0).length > 0", "true\n");
+}
+
+void ItemImageTests::saveGif()
+{
+    if ( !QImageReader::supportedImageFormats().contains("gif") )
+        SKIP("GIF image format is unsopported on this platform");
+
+    const auto data = QByteArray::fromBase64(
+        "R0lGODlhCAAIAPQAAAAAAE5WUURsU0ZuVVxkX3Jycnx8fFKCZFyRcFyScF6Ucl+Wc1+WdGmLdmuMeGGadmqqgmurg5OUlJubm56dnaGhobW1tb2+vgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACH5BAEAAAAAIf8LSW1hZ2VNYWdpY2sNZ2FtbWE9MC40NTQ1NQAsAAAAAAgACAAABSYgAAwiKSZLkjCJ+AhRJDzikTTOcohAIF0EHmBSsVCEhkLBIGwCQgA7");
+    TEST( m_test->setClipboard("TEST") );
+    WAIT_ON_OUTPUT("read(0)", "TEST");
+    TEST( m_test->setClipboard(data, "image/gif") );
+    WAIT_ON_OUTPUT("read('image/png', 0).length > 0", "true\n");
+}
+
+void ItemImageTests::saveWebp()
+{
+    if ( !QImageReader::supportedImageFormats().contains("webp") )
+        SKIP("WEBP image format is unsopported on this platform");
+
+    const auto data = QByteArray::fromBase64(
+        "UklGRrgAAABXRUJQVlA4WAoAAAAQAAAABwAABwAAQUxQSEEAAAAACFmTUFCTWQhN3url5ereTXLuwPn5wO5yLrvx/v7xuy4AIbD//7AhAAARw/z8wxEAAA2vqKivDAAAAEc2NkcAAABWUDggUAAAAPABAJ0BKggACAACADQljAJ0AQ8M/qJ2gAD+3i3bP+bdQx6F+ZyiDZNsPKvgXJjteWOHwsl4XY+Utm/aW5R74XTu3MMrnj1KVx5aDzswFkAA");
+    TEST( m_test->setClipboard("TEST") );
+    WAIT_ON_OUTPUT("read(0)", "TEST");
+    TEST( m_test->setClipboard(data, "image/webp") );
+    WAIT_ON_OUTPUT("read('image/png', 0).length > 0", "true\n");
+}

--- a/plugins/itemimage/tests/itemimagetests.h
+++ b/plugins/itemimage/tests/itemimagetests.h
@@ -1,0 +1,28 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+#pragma once
+
+#include "tests/testinterface.h"
+
+#include <QObject>
+
+class ItemImageTests final : public QObject
+{
+    Q_OBJECT
+public:
+    explicit ItemImageTests(const TestInterfacePtr &test, QObject *parent = nullptr);
+
+private slots:
+    void initTestCase();
+    void cleanupTestCase();
+    void init();
+    void cleanup();
+
+    void supportedFormats();
+    void savePng();
+    void saveBmp();
+    void saveGif();
+    void saveWebp();
+
+private:
+    TestInterfacePtr m_test;
+};

--- a/src/CMakeLists.txt
+++ b/src/CMakeLists.txt
@@ -193,26 +193,28 @@ else()
         COMPONENT Runtime
         )
 
-    set(FIXUP_BUNDLE_APPS "\${CMAKE_INSTALL_PREFIX}/${COPYQ_EXECUTABLE_NAME}.app")
-
-    get_property(_copyq_installed_plugins GLOBAL PROPERTY COPYQ_INSTALLED_PLUGINS)
-
-    # Directories to look for dependencies
-    set(FIXUP_BUNDLE_DEP_DIRS "${CMAKE_BINARY_DIR};${QT_LIBRARY_DIRS};${${copyq_qt}Widgets_DIR}/../..")
-
-    message(STATUS "Fixup app: ${FIXUP_BUNDLE_APPS}")
-    message(STATUS "Fixup app plugins: ${_copyq_installed_plugins}")
-    message(STATUS "Fixup qt plugins: ${FIXUP_BUNDLE_QT_PLUGINS}")
-    message(STATUS "Fixup dirs: ${FIXUP_BUNDLE_DEP_DIRS}")
-    install(CODE "
-        include(BundleUtilities)
-        set(BU_CHMOD_BUNDLE_ITEMS ON)
-        fixup_bundle(\"${FIXUP_BUNDLE_APPS}\" \"${FIXUP_BUNDLE_QT_PLUGINS};${_copyq_installed_plugins}\" \"${FIXUP_BUNDLE_DEP_DIRS}\")
-        verify_app(\"${FIXUP_BUNDLE_APPS}\")
-        "
-        COMPONENT Runtime
-        )
-
-    set(CPACK_GENERATOR "DragNDrop")
-    include(CPack)
+    # FIXME: Disable cpack for now due to unresolvable libraries:
+    #        https://github.com/Homebrew/homebrew-core/issues/140930
+    # set(FIXUP_BUNDLE_APPS "\${CMAKE_INSTALL_PREFIX}/${COPYQ_EXECUTABLE_NAME}.app")
+    #
+    # get_property(_copyq_installed_plugins GLOBAL PROPERTY COPYQ_INSTALLED_PLUGINS)
+    #
+    # # Directories to look for dependencies
+    # set(FIXUP_BUNDLE_DEP_DIRS "${CMAKE_BINARY_DIR};${QT_LIBRARY_DIRS};${QT_PLUGINS_DIR}/iconengines;${QT_PLUGINS_DIR}/imageformats;${QT_PLUGINS_DIR}/platforms;${QT_PLUGINS_DIR}/tls;${${copyq_qt}Widgets_DIR}/../..")
+    #
+    # message(STATUS "Fixup app: ${FIXUP_BUNDLE_APPS}")
+    # message(STATUS "Fixup app plugins: ${_copyq_installed_plugins}")
+    # message(STATUS "Fixup qt plugins: ${FIXUP_BUNDLE_QT_PLUGINS}")
+    # message(STATUS "Fixup dirs: ${FIXUP_BUNDLE_DEP_DIRS}")
+    # install(CODE "
+    #     include(BundleUtilities)
+    #     set(BU_CHMOD_BUNDLE_ITEMS ON)
+    #     fixup_bundle(\"${FIXUP_BUNDLE_APPS}\" \"${FIXUP_BUNDLE_QT_PLUGINS};${_copyq_installed_plugins}\" \"${FIXUP_BUNDLE_DEP_DIRS}\")
+    #     verify_app(\"${FIXUP_BUNDLE_APPS}\")
+    #     "
+    #     COMPONENT Runtime
+    #     )
+    #
+    # set(CPACK_GENERATOR "DragNDrop")
+    # include(CPack)
 endif()

--- a/src/tests/tests.cpp
+++ b/src/tests/tests.cpp
@@ -490,6 +490,10 @@ public:
 
     QByteArray verifyClipboard(const QByteArray &data, const QString &mime, bool exact = true) override
     {
+        // Due to image conversions in clipboard check only png.
+        if ( mime.startsWith(QStringLiteral("image/")) && mime != QStringLiteral("image/png") )
+            return verifyClipboard("PNG", QStringLiteral("image/png"), false);
+
         PerformanceTimer perf;
 
         SleepTimer t(5000);

--- a/utils/github/bundle-macos.sh
+++ b/utils/github/bundle-macos.sh
@@ -2,10 +2,18 @@
 # Creates macOS bundle.
 set -xeuo pipefail
 
-cpack
+app_bundle_path="CopyQ.app"
 
-app_bundle_path="$(echo _CPack_Packages/Darwin/DragNDrop/copyq-*/CopyQ.app)"
 executable="${PWD}/${app_bundle_path}/Contents/MacOS/CopyQ"
+plugins=("$app_bundle_path/Contents/PlugIns/copyq/"*.so)
+qt_bin="$(brew --prefix qt@5)/bin"
+
+rm -rf "$app_bundle_path/Contents/PlugIns/"{platforminputcontexts,printsupport,qmltooling}
+
+"$qt_bin/macdeployqt" "$app_bundle_path" -dmg -verbose=2 -always-overwrite -no-plugins \
+    "${plugins[@]/#/-executable=}"
+
+ls -Rl "$app_bundle_path"
 
 # Test the app before deployment.
 "$executable" --help
@@ -19,8 +27,8 @@ ls "$("$executable" info translations)/"
 test "$("$executable" info has-global-shortcuts)" -eq "1"
 
 # Uninstall local Qt to make sure we only use libraries from the bundle
-remove=$(brew list --formula)
-brew remove --force --ignore-dependencies $remove
+brew remove --ignore-dependencies --force \
+    qt@5 copyq/kde/kf5-knotifications freetype
 
 (
     export PATH=""
@@ -35,7 +43,6 @@ brew remove --force --ignore-dependencies $remove
 
 # Print dependencies to let us further make sure that we don't depend on local libraries
 otool -L "$executable"
+otool -L "$app_bundle_path/Contents/PlugIns/"*/*.dylib
 otool -L "$app_bundle_path/Contents/PlugIns/copyq/"*
 otool -L "$app_bundle_path/Contents/Frameworks/"Qt*.framework/Versions/5/Qt*
-
-mv copyq-*.dmg CopyQ.dmg

--- a/utils/github/install-linux.sh
+++ b/utils/github/install-linux.sh
@@ -49,6 +49,8 @@ else
 fi
 
 packages=(
+    ninja-build
+
     # fixes some issues on X11
     libxfixes-dev
     libxtst-dev

--- a/utils/github/install-macos.sh
+++ b/utils/github/install-macos.sh
@@ -23,4 +23,4 @@ rm -rf \
 
 brew tap copyq/kde utils/github/homebrew/
 
-brew install qt5 copyq/kde/kf5-knotifications
+brew install qt@5 copyq/kde/kf5-knotifications

